From 8ccb7329fa8be3537c42b3e2e36c6b80cd231ceb Mon Sep 17 00:00:00 2001
From: Yuan Juntao <juntao.yuan@intel.com>
Date: Fri, 8 May 2015 11:58:41 +0800
Subject: [PATCH] mmc: sdhci: disabling data irq if polling method is using for
 tuning

If host uses polling method to detect the finish of tuning process, data
interrupt signal should be disabled to prevent unwanted IRQ requests.

Change-Id: I2763d5a23d00c6115f785a32e259fcddeb948b00
Signed-off-by: Yuan Juntao <juntao.yuan@intel.com>
---
 drivers/mmc/host/sdhci.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/mmc/host/sdhci.c b/drivers/mmc/host/sdhci.c
index 6f59a657f485..7742d8a287ad 100644
--- a/drivers/mmc/host/sdhci.c
+++ b/drivers/mmc/host/sdhci.c
@@ -2040,6 +2040,16 @@ static int sdhci_execute_tuning(struct mmc_host *mmc, u32 opcode)
 	ier = sdhci_readl(host, SDHCI_INT_ENABLE);
 	sdhci_clear_set_irqs(host, ier, SDHCI_INT_DATA_AVAIL);
 
+	if (unlikely(host->quirks2 & SDHCI_QUIRK2_TUNING_POLL)) {
+		/*
+		 * Tuning poll doesn't need data interrupt signal.
+		 * Disable it to prevent unwanted irq requests.
+		 */
+		u32 no_data_int = sdhci_readl(host, SDHCI_SIGNAL_ENABLE);
+		no_data_int &= ~SDHCI_INT_DATA_AVAIL;
+		sdhci_writel(host, no_data_int, SDHCI_SIGNAL_ENABLE);
+	}
+
 	/*
 	 * Issue CMD19 repeatedly till Execute Tuning is set to 0 or the number
 	 * of loops reaches 40 times or a timeout of 150ms occurs.
-- 
1.9.1

